src/opf/dcp.jl

```diff
@@ -1,263 +1,171 @@
+    @variable(model, va[1:N])
 
-    # bus voltage angle
-    JuMP.@variable(model, va[1:N])
-
-    # generator active dispatch
-    JuMP.@variable(model, ref[:gen][g]["pmin"] <= pg[g in 1:G] <= ref[:gen][g]["pmax"])
-
-    # branch flows
-    JuMP.@variable(model, -ref[:branch][l]["rate_a"] <= pf[(l,i,j) in ref[:arcs_from]] <= ref[:branch][l]["rate_a"])
-
-    # https://github.com/lanl-ansi/PowerModels.jl/blob/d9c00d228a4e875ac3426425ad6c8f8338309efa/src/form/dcp.jl#L255-L258
-    # pf_expr = Dict{Any, Any}( ((l,i,j), pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from] )
-    # pf_expr = merge(pf_expr, Dict( ((l,j,i), -1.0*pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from] ))
-    pf_expr = [((l,i,j), pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from]]
-    pf_expr = vcat(pf_expr, [((l,j,i), -1.0*pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from]])
+    # Active and reactive dispatch
+    @variable(model, pg[g in 1:G])
 
-    model[:pf] = pf = JuMP.Containers.DenseAxisArray(collect(getfield.(pf_expr, 2)), collect(getfield.(pf_expr, 1)))
+    # Branch flows
+    @variable(model, pf[e in 1:E])
 
-    #
+    # 
     #   II. Constraints
     #
 
+    # Generation bounds (both zero if generator is off)
+    set_lower_bound.(pg, gen_status .* pgmin)
+    set_upper_bound.(pg, gen_status .* pgmax)
+
+    # Flow bounds (both zero if branch is off)
+    set_lower_bound.(pf, branch_status .* -smax)
+    set_upper_bound.(pf, branch_status .* smax)
+
     # Slack bus
-    JuMP.@constraint(model, slack_bus, va[i0] == 0.0)
+    @constraint(model, slack_bus, va[i0] == 0.0)
 
     # Nodal power balance
-    JuMP.@constraint(model,
+    @constraint(model,
         kirchhoff[i in 1:N],
-        sum(pg[g] for g in ref[:bus_gens][i])
-        - sum(pf[a] for a in ref[:bus_arcs][i]) 
-        ==
-        sum(load["pd"] for load in bus_loads[i])
-        + sum(shunt["gs"] for shunt in bus_shunts[i])
+        sum(pg[g] for g in bus_gens[i] if gen_status[g])
+        - sum(pf[a] for a in bus_arcs_fr[i] if branch_status[a])
+        - sum(-pf[a] for a in bus_arcs_to[i] if branch_status[a])
+        == 
+        pd[i] + gs[i]
     )
 
-    # Branch power flow physics and limit constraints
-    # We pre-allocate constraint containers for simplicity
-    model[:voltage_difference_limit] = Vector{JuMP.ConstraintRef}(undef, E)
-    model[:ohm_eq] = Vector{JuMP.ConstraintRef}(undef, E)
-    for (i,branch) in ref[:branch]
-        data["branch"]["$i"]["br_status"] == 0 && continue  # skip branch
-
-        f_idx = (i, branch["f_bus"], branch["t_bus"])
-
-        p_fr = pf[f_idx]
-
-        va_fr = va[branch["f_bus"]]
-        va_to = va[branch["t_bus"]]
+    model[:ohm] = Vector{ConstraintRef}(undef, E)
+    model[:va_diff] = Vector{ConstraintRef}(undef, E)
 
-        g, b = PM.calc_branch_y(branch)
+    for e in 1:E if branch_status[e]
+            # Branch power flow physics and limit constraints
+            model[:ohm][e] = @constraint(model, -b[e] * (va[bus_fr[e]] - va[bus_to[e]]) - pf[e] == 0)
```

Comment:
For Ohm's law, you can write it `pf[e] = branch_status[e] * b[e] * (va[j] - va[i])` (up to sign mistake).
This creates the usual constraint for active branches, and the constraint `pf[e] == 0` for inactive branches, without messing up the physics.
It will also ensure that Ohm's law constraints are defined for all branches

Since `pf` variables for inactive branches will not appear in any constraint, any corresponding dual should be zero.

Reply:
The potential for sign mistake is exactly why I put it in this form :) We actually had the `b[e] * (va[j]-va[i])` before, but I figured JuMP would simplify the double-negative for us.

I like the idea to multiply by `branch_status[e]` instead of the `for ... if ...`. I'll apply this to all the formulations. To be clear, power balance should stay as-is? There, we do a `pf[e] for e in bus_arcs if branch_status[e]` but we can do `branch_status[e] * pf[e] for e in bus_arcs` instead. I believe JuMP filters out any variables with zero coefficient:

```julia
julia> m = Model()
A JuMP Model
├ solver: none
├ objective_sense: FEASIBILITY_SENSE
├ num_variables: 0
├ num_constraints: 0
└ Names registered in the model: none

julia> @variable(m, x)
x

julia> @variable(m, y)
y

julia> @constraint(m, con, x+0y==1)
con : x = 1

julia> constraint_object(con)
ScalarConstraint{AffExpr, MathOptInterface.EqualTo{Float64}}(x, MathOptInterface.EqualTo{Float64}(1.0))
```

Reply:
Double negative: yes, that will be simplified for us. JuMP will also filter out zero coefficients, assuming we use non-direct mode, otherwise I think the solver might get a zero coefficient.

JuMP devs do discourage the use of `sum(x[i] for i in ... if <some_condition_on_i>)`, because it's a great way to get worse complexity.

Performance-wise, I'm not too worried with using either formulation, because we usually have a small number of inactive branches. Both will give mathematically equivalent constraints.